core/templates/dev/head/domain/topic/TopicObjectFactory.js

```diff
@@ -180,29 +190,12 @@ oppia.factory('TopicObjectFactory', ['SubtopicObjectFactory',
     };
 
     Topic.prototype.addUncategorizedSkillId = function(skillId) {
-      if (this._uncategorizedSkillIds.indexOf(skillId) !== -1) {
-        return false;
-      }
-      var skillIsPresentInSomeSubtopic = false;
-      for (var i = 0; i < this._subtopics.length; i++) {
-        var skillIds = this._subtopics[i].getSkillIds();
-        if (skillIds.indexOf(skillId) !== -1) {
-          skillIsPresentInSomeSubtopic = true;
-          break;
-        }
-      }
-      if (!skillIsPresentInSomeSubtopic) {
-        this._uncategorizedSkillIds.push(skillId);
-        return;
-      }
-      return false;
+      this._uncategorizedSkillIds.push(skillId);
```

Comment:
For these and some other functions, it's not a bad idea to add docstrings/comments saying "callers should ensure that the skill ID does not already occur anywhere in the topic", to make clear whose responsibility the validation is. Otherwise it looks a bit unsafe.

More generally, it might not be a bad idea to add JSDoc for most of these functions.

Reply:
Actually, I thought more about this:

> Done, moved all such checks to TopicUpdateService, since that's what the external controllers would be calling and so any errors can be found there itself, before the changelists are modified.

On further thought, it might actually be better to have the checks in the domain object. Let me explain the thinking and see if you agree?

The reason that we do validation checks in the controller (in the backend) is because every incoming request has to go through the controller layer. So catching errors there is sensible before it goes further down into the application. The idea is to minimize the surface you need to guard (almost like a min-cut in graph algorithms) in a way that makes it easy to have some assurance that you're catching everything.

If we apply that principle here, then it seems better to have the checks in the topic object factory, unless the topic update service is the only way these objects can be modified (and every developer knows that). Otherwise, if other things modify the Topic object, then they'll need to add their own validation checks as well. And, generally, since we're passing the Topic object all over the place, it doesn't seem unlikely that someone's going to try and do that in the future, since all its methods would be directly accessible.

So -- on thinking about it more, I lean towards having the validation checks in the Topic object instead. But what do you think of the above reasoning? I think it's worth debating since it's going to set a pattern for the other editors.

(And sorry for not having more of an opinion on it when I reviewed previously!)

Reply:
Yeah, I did think of that (about whether only TopicUpdateService would be accessing them), but in TopicUpdateService also, a check should be added as the new_value field used in the changelist is calculated separately (not using topic prototypes), so since a check has to be there anyways (for canonical and additional stories), I moved all of them there.

Reply:
Hm, I think you don't need those checks for the new_value field in TopicUpdateService. If (say) topic.addAdditionalStoryId() throws an error or returns false, is the change discarded? If you can verify that then I don't think you'll have any problems with validity of new_value since either both it and the change itself would be valid, or both would be invalid -- right?

Reply:
Done, changed all the 'return false' to throw an error. Also modified UndoRedoService to  first apply changes, then append to changelist (earlier, it was the other way around), so that if an error occurs in applying, changelist won't be made, and added tests to check this.